crates/bifrost/src/read_stream.rs

```diff
@@ -961,4 +976,97 @@ mod tests {
+                .enable_local_loglet(&config)
+                .enable_in_memory_loglet();
+            let bifrost = svc.handle();
+            svc.start().await.expect("loglet must start");
+
+            // prepare a chain that starts from Lsn 10 (we expect trim from OLDEST -> 9)
+            let old_version = bifrost.inner.metadata.logs_version();
+            let mut builder = bifrost.inner.metadata.logs().clone().into_builder();
+            let mut chain_builder = builder.chain(&LOG_ID).unwrap();
+            assert_eq!(1, chain_builder.num_segments());
+            let new_segment_params = new_single_node_loglet_params(ProviderKind::Local);
+            chain_builder.append_segment(Lsn::new(10), ProviderKind::Local, new_segment_params)?;
+            chain_builder.trim_prefix(Lsn::new(10));
+            assert_eq!(1, chain_builder.num_segments());
+            assert_eq!(Lsn::from(10), chain_builder.tail().base_lsn);
+
+            let new_metadata = builder.build();
+            let new_version = new_metadata.version();
+            assert_eq!(new_version, old_version.next());
+            node_env
+                .metadata_store_client
+                .put(
+                    BIFROST_CONFIG_KEY.clone(),
+                    new_metadata,
+                    restate_metadata_store::Precondition::MatchesVersion(old_version),
+                )
+                .await?;
+
+            // make sure we have updated metadata.
+            bifrost
+                .inner
+                .metadata
+                .sync(MetadataKind::Logs, TargetVersion::Latest)
+                .await?;
+
+            let mut reader = bifrost.create_reader(LOG_ID, Lsn::OLDEST, Lsn::MAX)?;
+
+            // append a few records
+            for i in 10..=13 {
+                let lsn = bifrost
+                    .append(LOG_ID, Payload::new(format!("record-{}", i)))
+                    .await?;
+                assert_eq!(Lsn::from(i), lsn);
+            }
+
+            // first read should be the trim gap.
+            let record = reader.next().await.expect("to stay alive")?;
+
+            assert_that!(
+                record,
+                pat!(LogRecord {
+                    offset: eq(Lsn::OLDEST),
+                    record: pat!(Record::TrimGap(pat!(TrimGap {
+                        to: eq(Lsn::from(9)),
+                    })))
+                })
+            );
+
+            assert_that!(reader.read_pointer(), eq(Lsn::from(10)));
+
+            // read records
+            for i in 10..=13 {
+                let record = reader.next().await.expect("to stay alive")?;
+                assert_eq!(Lsn::from(i), record.offset);
+                assert_eq!(
+                    Payload::new(format!("record-{}", i)).body(),
+                    record.record.into_payload_unchecked().body()
+                );
+            }
+
+            anyhow::Ok(())
+        })
+        .await?;
+        Ok(())
```

Comment:
Should we shutdown the `task_center` and `RocksDbManager`?

Reply:
Not needed as we don't spawn any tasks. In the future, this will be automatic via a drop handle for the environment (I hope).